### PR TITLE
More complete type stubs for tkinter Canvas methods.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This change

* Establishes a NewType called `_ObjectId`, meant to ensure that arbitrary handles aren't created.
* Defines more specific method declarations for `Canvas` instances.

### CODE REVIEW

**Comment by hauntsaninja** [comment]
It'll be a couple years before typeshed can use PEP 570 syntax, for now positional-only arguments are indicated by prefixing with a double underscore.

**Comment by srittau** [comment]
Except for some auto generate code for protobuf, this is the first time we use `NewType` in typeshed. Just something we need to watch out for after the change is published.

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:128]:
```suggestion
_ObjectId = NewType("_ObjectId", int)
```

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1121]: As hauntsaninja pointed out:

```suggestion
    def create_line(self, __x0: int, __y0: int, __x1: int, __y1: int,
```

(unfortunately)

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1145]:
```suggestion
    def create_oval(self, __x0: int, __y0: int, __x1: int, __y1: int,
```

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1170]:
```suggestion
    def create_polygon(self, __x0: int, __y0: int, __x1: int, __y1: int, *args: int,
```

**Comment by hauntsaninja** [comment]
Wow, I cannot wait to be done with Python 3.5.1 :-)

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1198]:
```suggestion
    def create_rectangle(self, __x0: int, __y0: int, __x1: int, __y1: int,
```

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1223]:
```suggestion
    def create_text(self, __x: int, __y: int,
```

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1239]:
```suggestion
    def create_window(self, __x: int, __y: int,
```

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1127]: Nit: Built-in generics are now mostly supported for typeshed.

```suggestion
        arrowshape: tuple[int, int, int] = ...,
```

(And a few more `Tuple` instances below.)

File: stdlib/tkinter/__init__.pyi
Comment (srittau) [inline stdlib/tkinter/__init__.pyi:1248]: Nit: Import `Iterable` from `collections.abc` and just use `Iterable` here.

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, remarks below! Also, running black on this file should hopefully fix the build issues.

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [reply stdlib/tkinter/__init__.pyi:128]: I would rename this to `_CanvasItemId`:
- There are lots of classes in this file, but only canvases use `int` IDs.
- The canvas docs say "item IDs".

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [reply stdlib/tkinter/__init__.pyi:1248]: Please use `_TkinterSequence`

Comment (DanielRosenwasser) [reply stdlib/tkinter/__init__.pyi:1248]: Hm. Should `_TkinterSequence` be redefined as this?

```diff
- _TkinterSequence = Union[List[_T], Tuple[_T, ...]]
+ _TkinterSequence = Union[List[_T], Iterable[_T, ...]]
```

**Comment by DanielRosenwasser** [comment]
I can't seem to get [black](https://pypi.org/project/black/) installed on Windows for whatever reason, so I will hold off on that step if that's acceptable to you all.

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [reply stdlib/tkinter/__init__.pyi]: No. Tkinter doesn't accept all iterables. There's a comment showing how it doesn't accept `collections.deque`.

**Comment by Akuli** [comment]
I blacked the code (pull before making more changes), but it is also possible on windows. I think something like:

```
py -m venv env
env\Scripts\activate.bat
pip install black
black stdlib
```

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [reply stdlib/tkinter/__init__.pyi]: While `_CanvasObjectId` is slightly better, I think `_CanvasItemId` would be even better, for consistency with [the docs](http://tcl.tk/man/tcl8.7/TkCmd/canvas.htm#M16).

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [inline stdlib/tkinter/__init__.pyi:1220]: A few things that also apply to other `create_foo()` methods:
- The coordinates can be `float`. Use `float` instead of `int`. There's no need to do `Union[float, int]`, because type checkers consider `int` to also be included when you say `float`.
- Use `*` to mark rest of the arguments keyword-only: `__y1: int, *, activedash: _Color = ...`

**Review by Akuli — CHANGES_REQUESTED** [review]

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [reply stdlib/tkinter/__init__.pyi]: There seems to be a problem with `tuple`, as just changing `Tuple` to `tuple` causes mypy primer failures unrelated to canvas. Maybe this is the same tuple problems as in #4820 ?

File: stdlib/tkinter/__init__.pyi
Comment (DanielRosenwasser) [reply stdlib/tkinter/__init__.pyi:1220]: Okay, I see my mistake. Both my original and current revisions were incorrect because they considered the latter arguments could be positional.

Comment (DanielRosenwasser) [reply stdlib/tkinter/__init__.pyi]: A lot of the coordinates are currently `Union[int, str]` - so I assume those should become `Union[float, str]`.

Comment (Akuli) [reply stdlib/tkinter/__init__.pyi]: You can also try setting the options to e.g. `"2c"`. If that works, the correct type is `_ScreenUnits`.

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [inline stdlib/tkinter/__init__.pyi:1144]: Usually `bool` is best for things like this. There's more discussion about this in #4669.

```suggestion
        smooth: bool = ...,
```

**Review by Akuli — CHANGES_REQUESTED** [review]

**Comment by Akuli** [comment]
Can you find/replace `tuple` back to `Tuple`? The lower-case `tuple` seems to be causing mypy problems (#4820 ). Once those are no longer a problem, we can make a separate pull request that changes `Tuple` to `tuple`.

File: stdlib/tkinter/__init__.pyi
Comment (DanielRosenwasser) [reply stdlib/tkinter/__init__.pyi]: Okay, it's just that `delete` definitely seems to operate on any iterable recursively.

**Comment by Akuli** [comment]
Often you can use a tag string instead of an item ID. The remaining mypy_primer failures are from code that does `canvas.delete('all')`, where `'all'` is a tag string. That should be accepted whenever [the manual page](https://www.tcl.tk/man/tcl8.7/TkCmd/canvas.htm) says `tagOrId`.

To fix this, you can replace `_CanvasObjectId` with `Union[str, _CanvasObjectId]` in a few places, or add a type alias:

```python
_TagOrId = Union[_CanvasObjectId, str]
```

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [reply stdlib/tkinter/__init__.pyi]: The source code is:

```
    def delete(self, *args):
        """Delete items identified by all tag or ids contained in ARGS."""
        self.tk.call((self._w, 'delete') + args)
```

So each item of `args` is passed to `self.tk.call`, which wants `_TkinterSequence` (but silently does the wrong thing with e.g. deques):

```
>>> import tkinter, collections
>>> c = tkinter.Canvas()
>>> c.pack()
>>> c.create_rectangle(100, 100, 200, 200)
1
>>> c.delete([1])   # rectangle goes away
>>> c.create_rectangle(100, 100, 200, 200)
2
>>> c.delete(collections.deque([2]))  # rectangle DOES NOT GO AWAY
```

Comment (Akuli) [reply stdlib/tkinter/__init__.pyi]: Actually, `c.delete([3, 4])` doesn't work either, so we shouldn't use `_TkinterSequence` or `Iterable` at all.

**Comment by Akuli** [comment]
@DanielRosenwasser Are you still interested in working on this? There's no hurry, but I can do the rest if you want.

**Comment by DanielRosenwasser** [comment]
Hey @akuli let me know if there's anything I've missed in the latest changes.

**Comment by Akuli** [comment]
The `mypy_primer` failure comes from code that (I wrote and) uses `int` as the type for canvas items. What would be the recommended way once this PR is merged? We have a few options:
- Try to have mypy automatically determine the type (not possible in this case)
- Use `tkinter._CanvasItemId` (I don't think this works, and it's fragile and not discoverable, not guessable, not documented anywhere)
- Rename it to `CanvasItemId` and use that (would work, but still not discoverable/guessable/documented)
- Just get rid of `NewType` and everything will work

What do you think?

**Comment by DanielRosenwasser** [comment]
I've been thinking about this a bit and it really depends on how much someone wants to protect against breaking the abstraction in addition to making sure you can't forge a fake object ID (on purpose or accidentally). In either case, it's not that important and unlikely that an ID will ever stop being an int.

**Comment by DanielRosenwasser** [comment]
So I'll try to just turn it into a plain type alias to int later today.

File: stdlib/tkinter/__init__.pyi
Comment (Akuli) [inline stdlib/tkinter/__init__.pyi:13]:
```suggestion
```

**Review by Akuli — CHANGES_REQUESTED** [review]

**Comment by Akuli** [comment]
The code needs `black`ing again. Let me know whether you want to black it yourself or you want me to black it again.

**Comment by Akuli** [comment]
I decided to `black` it

**Review by Akuli — APPROVED** [review]
Thanks!

**Comment by DanielRosenwasser** [comment]
Ah just as I was figuring out how to `venv` 😄

Thanks!

**Pull request merged**
